src/rt/rwlock.rs

```diff
@@ -0,0 +1,269 @@
+        );
+    }
+
+    pub(crate) fn try_acquire_read_lock(&self) -> bool {
+        self.obj.branch_opaque();
+        self.post_acquire_read_lock()
+    }
+
+    pub(crate) fn release_read_lock(&self) {
+        super::execution(|execution| {
+            let state = self.get_state(&mut execution.objects);
+
+            state.lock = None;
+
+            state
+                .synchronize
+                .sync_store(&mut execution.threads, Release);
+
+            if state.seq_cst {
+                execution.threads.seq_cst();
+            }
+
+            let thread_id = execution.threads.active_id();
+
+            for (id, thread) in execution.threads.iter_mut() {
+                if id == thread_id {
+                    continue;
+                }
+
+                let obj = thread
+                    .operation
+                    .as_ref()
+                    .map(|operation| operation.object());
+
+                if obj == Some(self.obj) {
+                    thread.set_runnable();
+                }
+            }
+        });
+    }
+
+    fn post_acquire_read_lock(&self) -> bool {
+        super::execution(|execution| {
+            let state = self.get_state(&mut execution.objects);
+            let thread_id = execution.threads.active_id();
+
+            let thread_set = match &state.lock {
+                None => {
+                    let mut threads: HashSet<thread::Id> = HashSet::new();
+                    threads.insert(thread_id);
+                    threads
+                }
+                Some(RwLockType::Read(current)) => {
+                    let mut threads: HashSet<thread::Id> = HashSet::new();
+                    threads.extend(current);
+                    threads.insert(thread_id);
+                    threads
+                }
+                Some(RwLockType::Write(_)) => return false,
+            };
+
+            dbg!(state.synchronize.sync_load(&mut execution.threads, Acquire));
+
+            if state.seq_cst {
+                execution.threads.seq_cst();
+            }
+
+            // Block all writer threads from attempting to acquire the RwLock
+            for (id, thread) in execution.threads.iter_mut() {
+                if thread_set.contains(&id) {
+                    continue;
+                }
+
+                let obj = thread
+                    .operation
+                    .as_ref()
+                    .map(|operation| operation.object());
+
+                if obj == Some(self.obj) {
+                    thread.set_blocked();
+                }
+            }
+
+            state.lock = Some(RwLockType::Read(thread_set));
+
+            true
+        })
+    }
+
+    fn is_read_locked(&self) -> bool {
+        super::execution(
+            |execution| match self.get_state(&mut execution.objects).lock {
+                Some(RwLockType::Read(_)) => true,
+                _ => false,
+            },
+        )
+    }
+}
+
+impl RwLock {
```

Comment:
Why are there multiple `impl` blocks here? Can these be combined into one impl block?

Reply:
It was just for readability. One block is the reading methods, one block is the writing methods, third for the rest. I can combine them 😄

Reply:
Oh, I was also asking about that in another commit.